graphql/schemabuilder/batch.go

```diff
@@ -0,0 +1,300 @@
+package schemabuilder
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/samsarahq/thunder/graphql"
+)
+
+// buildBatchFunction corresponds to buildFunction for a batchFieldFunc
+func (sb *schemaBuilder) buildBatchFunction(typ reflect.Type, m *method) (*graphql.Field, error) {
+	funcCtx := &batchFuncContext{parentTyp: typ}
+
+	if typ.Kind() == reflect.Ptr {
+		return nil, fmt.Errorf("source-type of buildBatchFunction cannot be a pointer (got: %v)", typ)
+	}
+
+	callableFunc, err := funcCtx.getFuncVal(m)
+	if err != nil {
+		return nil, err
+	}
+
+	in := funcCtx.getFuncInputTypes()
+	if len(in) == 0 {
+		return nil, fmt.Errorf("batch Field funcs require at least one input field")
+	}
+
+	in = funcCtx.consumeContext(in)
+	in, err = funcCtx.consumeRequiredSourceBatch(in)
+	if err != nil {
+		return nil, err
+	}
+	argParser, args, in, err := funcCtx.consumeArgs(sb, in)
+	if err != nil {
+		return nil, err
+	}
+	in = funcCtx.consumeSelectionSet(in)
+
+	// We have succeeded if no arguments remain.
+	if len(in) != 0 {
+		return nil, fmt.Errorf("%s arguments should be [context,]map[int][*]%s[, args][, selectionSet]", funcCtx.funcType, typ)
+	}
+
+	out := funcCtx.getFuncOutputTypes()
+	retType, out, err := funcCtx.consumeReturnValue(m, sb, out)
+	if err != nil {
+		return nil, err
+	}
+	out = funcCtx.consumeReturnError(out)
+	if len(out) > 0 {
+		return nil, fmt.Errorf("%s return should be [map[int]<Type>][,error]", funcCtx.funcType)
+	}
+
+	return &graphql.Field{
+		Resolve: func(ctx context.Context, source, funcRawArgs interface{}, selectionSet *graphql.SelectionSet) (interface{}, error) {
+			// TODO use real batching.
+			sources := []interface{}{source}
+			// Set up function arguments.
+			funcInputArgs := funcCtx.prepareResolveArgs(sources, funcRawArgs, ctx, selectionSet)
+
+			// Call the function.
+			funcOutputArgs := callableFunc.Call(funcInputArgs)
+
+			results, err := funcCtx.extractResultsAndErr(len(sources), funcOutputArgs, retType)
+			if err != nil {
+				return nil, err
+			}
+			return results[0], nil
+		},
+		Args:           args,
+		Type:           retType,
+		ParseArguments: argParser.Parse,
+		Expensive:      funcCtx.hasContext,
+	}, nil
+}
+
+// funcContext is used to parse the function signature in buildFunction.
+type batchFuncContext struct {
+	hasContext      bool
+	hasSource       bool
+	hasArgs         bool
+	hasSelectionSet bool
+	hasRet          bool
+	hasError        bool
+
+	funcType     reflect.Type
+	batchMapType reflect.Type
+	isPtrFunc    bool
+	parentTyp    reflect.Type
+}
+
+// getFuncVal returns a reflect.Value of an executable function.
+func (funcCtx *batchFuncContext) getFuncVal(m *method) (reflect.Value, error) {
```

Comment:
I find it confusing that this function both extracts something out of m and sticks on the funcCtx, and then returns something else.

I would inline functions this short in the caller.